### PR TITLE
Add rtl support

### PR DESCRIPTION
**Pull request by hosseinkhaledi**
We add full rtl support to angular material except material.components.slider
All components tested( except material.components.slider) and work correctly in rtl and ltr mode

### CODE REVIEW

**Comment by ThomasBurleson** [comment]
Please:

*  squash your commits, and 
*  make sure you have done a `pull --rebase origin master` (where origin is **https://github.com/angular/material/**).

File: src/core/style/mixins.scss [hunk 1]
```diff
@@ -105,7 +105,15 @@
...
-  #{$prop}: $value;
+  html[dir=ltr] & {
+    #{$prop}: $value;
+  }
+  body[dir=ltr] & {
+    #{$prop}: $value;
+  }
+  html:not([dir]) body:not([dir]) & {
+    #{$prop}: $value;
+  }
```
Comment (EladBezalel) [inline src/core/style/mixins.scss]: Why all of this is necessary if all of them does the same thing?

File: src/components/chips/chips.scss [hunk 2]
```diff
@@ -25,7 +25,8 @@ $contact-chip-name-width: rem(12) !default;
       .md-contact-name {
         display: inline-block;
         height: $chip-height;
-        margin-left: rem(0.8);
+        @include rtl(margin-left, rem(0.8), null);
+        @include rtl(margin-right, null, rem(0.8));
```
Comment (EladBezalel) [inline src/components/chips/chips.scss]: Maybe there's a way to make it smarter?

File: src/core/style/mixins.scss (hunk 1 above)
Comment (hosseinkhaledi) [inline src/core/style/mixins.scss]: line 108 is to handle html[dir=ltr]
line 111 is to handle body[dir=ltr]
line 114  is to handle case <html><body>... (html and body has not dir) so must be ltr
if do not this we can not reverse ltr to rtl simply for example
margin-left: rem(0.8);
must be 
 @include rtl(margin-left, rem(0.8), null);
@include rtl(margin-right, default_value, rem(0.8));
that default_value depends on other css rule and must be calculated

File: src/components/chips/chips.scss (hunk 2 above)
Comment (hosseinkhaledi) [inline src/components/chips/chips.scss]: if define rtl with 4 parameter like rlt2(margin-left, rem(0.8), margin-right, rem(0.8))

File: src/core/style/mixins.scss (hunk 1 above)
Comment (EladBezalel) [inline src/core/style/mixins.scss]: so maybe 
``` css
html[dir=ltr] &, body[dir=ltr] &, html:not([dir]) body:not([dir]) & {
  #{$prop}: $value;
}
```

**Comment by ThomasBurleson** [comment]
@hosseinkhaledi - this is looking very nice. Can you rebase with latest code from master and then make sure the tests pass. Thx.

**Comment by hosseinkhaledi** [comment]
@ThomasBurleson I applied the rebase action

**Comment by EladBezalel** [comment]
Closed with SHA 08bf4f9 #7245

**Pull request closed without merging**